internal/helpers/deployment.go

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package helpers
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	AnnotationPreDeleteHookStarted       = "vso.secrets.hashicorp.com/pre-delete-hook-started"
+	AnnotationInMemoryVaultTokensRevoked = "vso.secrets.hashicorp.com/in-memory-vault-tokens-revoked"
+	LabelSelectorControlPlane            = "control-plane=controller-manager"
+	StringTrue                           = "true"
+)
+
+func AwaitInMemoryVaultTokensRevoked(ctx context.Context, logger logr.Logger, c client.Client) {
+	selector, err := labels.Parse(LabelSelectorControlPlane)
+	if err != nil {
+		logger.Error(err, "failed to parse label selector", "selector", LabelSelectorControlPlane)
+		return
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			logger.Error(ctx.Err(), "failed to await in-memory vault tokens revoked")
+			return
+		default:
+			var list corev1.PodList
+			err = c.List(ctx, &list, client.MatchingLabelsSelector{
+				Selector: selector,
+			})
+			if err != nil {
+				logger.Error(err, "failed to get pod list", "selector", LabelSelectorControlPlane)
+			} else {
+				for _, pod := range list.Items {
+					if value, ok := pod.Annotations[AnnotationInMemoryVaultTokensRevoked]; ok && value == StringTrue {
+						logger.Info("Operator pods annotations updated", AnnotationInMemoryVaultTokensRevoked, StringTrue)
+						return
+					}
+				}
+			}
+			time.Sleep(300 * time.Millisecond)
+		}
+	}
+}
+
+func AwaitPreDeleteStarted(ctx context.Context, handler context.CancelFunc, logger logr.Logger) {
+	for {
```

Comment:
Per our DM, we decided to stick with the loop polling file instead of the `preStop` command for now due to the following reasons 
1. We use a distroless build for the manager container. `kill` won’t be available on distroless https://github.com/hashicorp/vault-secrets-operator/blob/b65708ca82157dd4102dea341aa6647f3cd43b5e/Dockerfile#L38
2. The only thing we can do is run the operator binary, and that could implement sending the HUP signal (for example, using `client-go` to remote exec a command like `vault-secrets-operator --start-shutdown`) to the original operator process. However, it's an anti-pattern to remote exec a command in the operator pod from a pre-delete hook pod.
3.  `preStop` is not applicable to the current design. The operator pod waits for the pre-delete hook to start before revoking the tokens, and the pre-delete hook waits for the operator pod's in-memory token revocation to finish.  `preStop` runs as the operator pod is getting shutdown. The “stopping” process only starts when the pod receives a deletion timestamp, which is after the pre-delete hook finishes.